crates/bevy_reflect/src/type_registry.rs

```diff
@@ -100,24 +107,57 @@ impl TypeRegistry {
         registry
     }
 
-    /// Registers the type `T`, adding reflect data as specified in the [`Reflect`] derive:
+    /// Attempts to register the type `T` if it has not yet been registered already.
+    ///
+    /// If the registration for type `T` already exists, it will not be registered again.
+    /// To register the type, overwriting any existing registration, use [register](Self::register) instead.
+    ///
+    /// Additionally, this will add the reflect [data](TypeData) as specified in the [`Reflect`] derive:
     /// ```ignore (Neither bevy_ecs nor serde "derive" are available.)
-    /// #[derive(Component, serde::Serialize, serde::Deserialize, Reflect)]
+    /// #[derive(Reflect)]
     /// #[reflect(Component, Serialize, Deserialize)] // will register ReflectComponent, ReflectSerialize, ReflectDeserialize
+    /// struct Foo;
     /// ```
     pub fn register<T>(&mut self)
     where
         T: GetTypeRegistration,
     {
-        self.add_registration(T::get_type_registration());
+        if self.contains(TypeId::of::<T>()) {
+            return;
+        }
+
+        self.overwrite_registration(T::get_type_registration());
+        T::register_type_dependencies(self);
     }
 
-    /// Registers the type described by `registration`.
-    pub fn add_registration(&mut self, registration: TypeRegistration) {
-        if self.registrations.contains_key(&registration.type_id()) {
-            return;
+    /// Attempts to register the type described by `registration`.
+    ///
+    /// If the registration for the type already exists, it will not be registered again.
+    ///
+    /// To forcibly register the type, overwriting any existing registration, use the
+    /// [`overwrite_registration`](Self::overwrite_registration) method instead.
+    ///
+    /// Returns `None` if the registration was successfully added,
+    /// or `Some(existing)` if it already exists.
+    pub fn add_registration(
+        &mut self,
+        registration: TypeRegistration,
+    ) -> Option<&TypeRegistration> {
+        if self.contains(registration.type_id()) {
+            self.get(registration.type_id())
+        } else {
+            self.overwrite_registration(registration);
```

Comment:
```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => {
                 let (_, old) = entry.replace_entry(registration);
                 Some(old)
            }
            Entry::Vacant(entry) => {
                 entry.insert(registration);
                 None
            }
        }
```
@MrGVSV this should work (sans formatting and typos).

Reply:
Hm, I think the problem with this is that we don't want to replace the old value. This function shouldn't overwrite.

Reply:
If that's the case, just flip the branches:

```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => None,
            Entry::Vacant(entry) => Some(&*entry.insert(registration),
```

Reply:
I think this is still flipped. When the entry is `Vacant`, we want to insert the registration and return `None`. When it's `Occupied` we don't want to insert the registration, and instead return `Some(existing)`.

Reply:
This should work then.

```suggestion
        match self.registrations.entry(registration.type_id()) {
            Entry::Occupied(entry) => Some(entry.get()),
            Entry::Vacant(entry) => {
                 entry.insert(registration);
                 None
            },
```

Reply:
I’m not at my computer right now but I think when I tried that it gave me an error about returning a reference to a local variable (`entry`). Is it giving it for you as well?

Reply:
Oh I didn't realize that a `get` takes a reference, you may need to use `into_mut` and then downgrade it to a read-only reference.

Reply:
Okay so this works, but I forgot that it actually needs to call `self.overwrite_registration` to compute everything and not just `entry.insert`, which means we run into borrow issues again 😵‍💫

Reply:
Ah nevermind then. This can be handled in a separate PR.

Reply:
Given that my only major concern with this approach has been startup performance for a method that will get called hundreds-to-thousands of times, I think it is worth frontloading this refactor. I've queued up some changes.